File: Euler-1D/input_VegaE.py
import numpy as np

def Cstar(T1, k, Rgas):
    return np.sqrt(Rgas * T1 / k) * ((k + 1.0) * 0.5) ** ((k + 1.0) / (2.0 * (k - 1.0)))

def nozzle_mdot_from_p(p1, At, cstar):
    return p1 * At / cstar

def area_mach_function(M, k):
    return (1.0 / M) * (((1.0 + 0.5 * (k - 1.0) * M * M)/(0.5 * (k + 1))) ** ((k + 1.0) / (2.0 * (k - 1.0))))

def solve_exit_mach(area_ratio, k):
    # Simple Newton-Raphson with clamp; start at M=2.5
    M = max(1.5, 0.5 * area_ratio + 1.0)
    # M = 3
    for _ in range(50):
        f = area_mach_function(M, k) - area_ratio
        # Numerical derivative
        dM = 1e-5 * max(1.0, M)
        df = (area_mach_function(M + dM, k) - area_mach_function(M - dM, k)) / (2.0 * dM)
        step = -f / df
        M = max(1.01, M + step)
        if abs(step) < 1e-8:
            break
    return M

def solve_area_mach(area_ratio, k_local, supersonic):
        if area_ratio <= 1.0:
            return 1.0

        if supersonic:
            lo = 1.0 + 1e-8
            hi = max(2.0, 0.5 * area_ratio + 1.0)
            while area_mach_function(hi, k_local) < area_ratio and hi < 100.0:
                hi *= 1.5
        else:
            lo = 1e-8
            hi = 1.0 - 1e-8

        for _ in range(80):
            mid = 0.5 * (lo + hi)
            f_mid = area_mach_function(mid, k_local) - area_ratio
            if supersonic:
                if f_mid > 0.0:
                    hi = mid
                else:
                    lo = mid
            else:
                if f_mid > 0.0:
                    lo = mid
                else:
                    hi = mid
        # print(lo, hi)
        return 0.5 * (lo + hi)


def nozzle_perf(p1, At, eps, k, Rgas, T1, p_amb):
    cstar = Cstar(T1, k, Rgas)
    mdot = nozzle_mdot_from_p(p1, At, cstar)
    A2 = eps * At
    M2 = solve_exit_mach(eps, k)
    T2 = T1 / (1.0 + 0.5 * (k - 1.0) * M2 * M2)
    p2 = p1 * (T2 / T1) ** (k / (k - 1.0))
    a_2 = np.sqrt(k * Rgas * T2)
    v_e = M2 * a_2
    F = mdot * v_e + (p2 - p_amb) * A2
    return F, mdot, v_e, p2/p1


def alpha_er_from_properties(cp_gas, mu_gas, K, cs_solid, rho_p, T1, Ts, T2, Tp):
    cp_gas = cp_gas*0.239/1000
    cs_solid = cs_solid*0.239/1000
    Pr = mu_gas * cp_gas * 4184 / K
    if (T2 - Tp) == 0:
        return 0.0
    return 0.0288 * cp_gas * (mu_gas ** 0.2) * (Pr ** (-2.0 / 3.0)) / (rho_p * cs_solid) * ((T1 - Ts) / (T2 - Tp))

# Erosive burning parameters
Tp0 = 300.0             # K, initial propellant temperature
T2prop = 1825.0         # K, propellant surface temperature at ignition
TSurf = 900.0           # K, ignition threshold at surface
T1prop = 3000.0         # K, combustion gas temperature for erosive burning calculation
# Ksolid = 0.2            # W/m-K, solid propellant thermal conductivity
Ksolid = 0.1            # W/m-K, solid propellant thermal conductivity
cpsolid = 1230.0        # J/kg-K, solid propellant specific heat

arb = 1e-4
nrb = 0.4
rhosolid = 1810.0       # kg/m^3

mugas = 9e-5            # kg/m-s
Kgas = 0.587            # W/m-K
cpgas = 2361            # J/kg-K, gas specific heat
kgas = 1.1487
Rgas = cpgas * (kgas-1) / kgas # J/kg-K
hreaction = 7e6


# Igniter parameters
kigniter = 1.22            # gamma for igniter gas, for ignition thermal state calculation 
cpgas_igniter = 1900       # J/kg-K, for ignition thermal state calculation
Rgas_igniter = cpgas_igniter * (kigniter-1) / kigniter
mugas_igniter = 8e-5         # kg/m-s, for ignition thermal state calculation    
Kgas_igniter = 0.4       # W/m-K, for ignition thermal state calculation
mig = 179.0                  # kg/s for 0.35s
vinj = 1500.0                   # m/s
hig = 3.5e6                 # J/kg

# Start values
pamb = 101325 # Pa
p0 = 101325 # Pa
rho0 = 1.225 # kg/m^3

# Nozzle and case geometry
Lcase = 8.69  # m
Rcase = 1.45 # m
Vfree = 8.58 # m^3
Ainitial = 1 # m^2
Pinitial = 2*np.sqrt(Ainitial*np.pi)
A2nozzle = np.pi*1.76*1.76/4 # m^2
epsnozzle = 16.0

# Case Geometry
# Lcase = 8.69  # m
# Rcase = 1.45 # m
# Vfree = 8.58 # m^3
# Ainitial = 1 # m^2

# # Nozzle
# A2nozzle = np.pi*1.76*1.76/4 # m^2
# epsnozzle = 16.0
# pamb = 101325 # Pa
# p0 = 101325 # Pa
# rho0 = 1.225 # kg/m^3

# Propellant
# rhoprop = 1810.0 # kg/m^3
# arb = 3e-4
# nrb = 0.4
# hreaction = 3.5e6  # reaction heat release for source term example

# Combustion gas
# k_gas = 1.1487
# # Rgas = 274.7249 # J/kg-K
# T1 = 3000.0 # K
# cpgas = 2361 # J/kg-K
# Rgas = cpgas * (k_gas-1) / k_gas # J/kg-K

# Igniter properties
# mig = 179.0 # kg/s for 0.35s
# vinj = 1500.0 # m/s
# hig = 3.5e6 # J/kg

# Erosive burning properties
# cpgas=2361*0.239/1000,    # kJ/kg-K to kcal/kg-K
# mu_gas=9e-5, # kg/m-s
# K=0.587,  # W/m-K+
# cs_solid=1230.0*0.239/1000,  # kJ/kg-K to kcal/kg-K
# rho_p=rhoprop, # kg/m^3

# TS=1060.0, # K
# T2=1825.0, # K
# Tp0=300.0, # K



# Erosive (Lenoir–Robillard)
alpha_er = alpha_er_from_properties(
    cp_gas=cpgas*0.239/1000,    # kJ/kg-K to kcal/kg-K
    mu_gas=mugas, # kg/m-s
    K=Kgas,  # W/m-K+
    cs_solid=cpsolid*0.239/1000,  # kJ/kg-K to kcal/kg-K
    rho_p=rhosolid, # kg/m^3
    T1=T1prop, # K
    Ts=TSurf, # K
    T2=1825.0, # K
    Tp=300.0, # K
)
beta_er = 53.0
# print(alpha_er)

File: Euler-1D-C/input_vegaE.c
#include "input_vegaE.h"

#include <math.h>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

const double Tp0 = 300.0;
const double T2prop = 1825.0;
const double TSurf = 900.0;
const double T1prop = 3000.0;
const double Ksolid = 0.1;
const double cpsolid = 1230.0;

const double arb = 1e-4;
const double nrb = 0.4;
const double rhosolid = 1810.0;

const double mugas = 9e-5;
const double Kgas = 0.587;
const double cpgas = 2361.0;
const double kgas = 1.1487;
const double Rgas = 2361.0 * (1.1487 - 1.0) / 1.1487;
const double hreaction = 7e6;

const double kigniter = 1.22;
const double cpgas_igniter = 1900.0;
const double Rgas_igniter = 1900.0 * (1.22 - 1.0) / 1.22;
const double mugas_igniter = 8e-5;
const double Kgas_igniter = 0.4;
const double mig = 179.0;
const double vinj = 1500.0;
const double hig = 3.5e6;

const double pamb = 101325.0;
const double p0 = 101325.0;
const double rho0 = 1.225;

const double Lcase = 8.69;
const double Rcase = 1.45;
const double Vfree = 8.58;
const double Ainitial = 1.0;
const double Pinitial = 2.0 * sqrt(1.0 * M_PI);
const double A2nozzle = M_PI * 1.76 * 1.76 / 4.0;
const double epsnozzle = 16.0;

double Cstar(double T1, double k, double Rg) {
    return sqrt(Rg * T1 / k) * pow((k + 1.0) * 0.5, (k + 1.0) / (2.0 * (k - 1.0)));
}

double nozzle_mdot_from_p(double p1, double At, double cstar) {
    return p1 * At / cstar;
}

double area_mach_function(double M, double k) {
    return (1.0 / M) * pow((1.0 + 0.5 * (k - 1.0) * M * M) / (0.5 * (k + 1.0)), (k + 1.0) / (2.0 * (k - 1.0)));
}

double solve_exit_mach(double area_ratio, double k) {
    double M = fmax(1.5, 0.5 * area_ratio + 1.0);
    for (int it = 0; it < 50; ++it) {
        double f = area_mach_function(M, k) - area_ratio;
        double dM = 1e-5 * fmax(1.0, M);
        double df = (area_mach_function(M + dM, k) - area_mach_function(M - dM, k)) / (2.0 * dM);
        double step = -f / df;
        M = fmax(1.01, M + step);
        if (fabs(step) < 1e-8) {
            break;
        }
    }
    return M;
}

double solve_area_mach(double area_ratio, double k_local, int supersonic) {
    if (area_ratio <= 1.0) {
        return 1.0;
    }

    double lo;
    double hi;
    if (supersonic) {
        lo = 1.0 + 1e-8;
        hi = fmax(2.0, 0.5 * area_ratio + 1.0);
        while (area_mach_function(hi, k_local) < area_ratio && hi < 100.0) {
            hi *= 1.5;
        }
    } else {
        lo = 1e-8;
        hi = 1.0 - 1e-8;
    }

    for (int it = 0; it < 80; ++it) {
        double mid = 0.5 * (lo + hi);
        double f_mid = area_mach_function(mid, k_local) - area_ratio;
        if (supersonic) {
            if (f_mid > 0.0) {
                hi = mid;
            } else {
                lo = mid;
            }
        } else {
            if (f_mid > 0.0) {
                lo = mid;
            } else {
                hi = mid;
            }
        }
    }
    return 0.5 * (lo + hi);
}

void nozzle_perf(double p1, double At, double eps, double k, double Rg, double T1, double p_amb,
                 double* F, double* mdot, double* v_e, double* p2_over_p1) {
    double cstar = Cstar(T1, k, Rg);
    double md = nozzle_mdot_from_p(p1, At, cstar);
    double A2 = eps * At;
    double M2 = solve_exit_mach(eps, k);
    double T2 = T1 / (1.0 + 0.5 * (k - 1.0) * M2 * M2);
    double p2 = p1 * pow(T2 / T1, k / (k - 1.0));
    double a2 = sqrt(k * Rg * T2);
    double ve = M2 * a2;
    *F = md * ve + (p2 - p_amb) * A2;
    *mdot = md;
    *v_e = ve;
    *p2_over_p1 = p2 / p1;
}

double alpha_er_from_properties(double cp_gas, double mu_gas, double K, double cs_solid,
                                double rho_p, double T1, double Ts, double T2, double Tp) {
    cp_gas = cp_gas * 0.239 / 1000.0;
    cs_solid = cs_solid * 0.239 / 1000.0;
    double Pr = mu_gas * cp_gas * 4184.0 / K;
    if ((T2 - Tp) == 0.0) {
        return 0.0;
    }
    return 0.0288 * cp_gas * pow(mu_gas, 0.2) * pow(Pr, -2.0 / 3.0) / (rho_p * cs_solid) * ((T1 - Ts) / (T2 - Tp));
}

const double alpha_er =
    0.0288 * ((2361.0 * 0.239 / 1000.0) * 0.239 / 1000.0) * pow(9e-5, 0.2) *
    pow((9e-5) * ((2361.0 * 0.239 / 1000.0) * 0.239 / 1000.0) * 4184.0 / 0.587, -2.0 / 3.0) /
    (1810.0 * ((1230.0 * 0.239 / 1000.0) * 0.239 / 1000.0)) * ((3000.0 - 900.0) / (1825.0 - 300.0));

const double beta_er = 53.0;
